File: packages/nitro-math/android/src/main/java/com/margelo/nitro/math/HybridMath.kt
package com.margelo.nitro.math

import com.margelo.nitro.core.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class HybridMath : HybridMathSpec() {
  override val pi: Double
    get() = Math.PI

  override fun add(a: Double, b: Double): Double {
    return a + b
  }

  override fun fetchScore(userId: String): Promise<Double> {
    if (userId.isEmpty()) {
      throw RuntimeException("userId cannot be empty")
    }

    return Promise.async {
      delay(1000)
      (0..100).random().toDouble()
    }
  }

  override fun startWork(onProgress: (Double) -> Unit) {
    CoroutineScope(Dispatchers.Default).launch {
      for (i in 1..10) {
        delay(50)
        onProgress(i / 10.0)
      }
    }
  }
}

File: packages/nitro-math/ios/HybridMath.swift
import NitroModules
import Foundation

class HybridMath: HybridMathSpec {
  var pi: Double {
    return Double.pi
  }

  func add(a: Double, b: Double) throws -> Double {
    return a + b
  }

  func fetchScore(userId: String) throws -> Promise<Double> {
    if userId.isEmpty {
      throw RuntimeError.error(withMessage: "userId cannot be empty")
    }

    return Promise.async {
      try await Task.sleep(nanoseconds: 1_000_000_000)
      return Double(Int.random(in: 0...100))
    }
  }

  func startWork(onProgress: @escaping (Double) -> Void) throws {
    Task {
      for i in 1...10 {
        try await Task.sleep(nanoseconds: 50_000_000)
        onProgress(Double(i) / 10.0)
      }
    }
  }
}
